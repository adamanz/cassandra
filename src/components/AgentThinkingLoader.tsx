import React from 'react';
import { motion } from 'framer-motion';
import { cn } from '@/utils/cn';

interface AgentThinkingLoaderProps {
  className?: string;
}

export const AgentThinkingLoader: React.FC<AgentThinkingLoaderProps> = ({ className }) => {
  return (
    <div className={cn('flex items-center space-x-2', className)}>
      <motion.div
        className="flex space-x-1"
        initial={{ opacity: 0 }}
        animate={{ opacity: 1 }}
        transition={{ duration: 0.5 }}
      >
        {[0, 1, 2].map((index) => (
          <motion.div
            key={index}
            className="w-2 h-2 bg-primary rounded-full"
            animate={{
              y: [0, -8, 0],
            }}
            transition={{
              duration: 0.6,
              repeat: Infinity,
              delay: index * 0.15,
              ease: "easeInOut",
            }}
          />
        ))}
      </motion.div>
      <span className="text-sm text-muted animate-pulse">Thinking...</span>
    </div>
  );
};

AgentThinkingLoader.displayName = 'AgentThinkingLoader';