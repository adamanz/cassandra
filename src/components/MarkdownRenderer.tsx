import React, { memo } from 'react';
import ReactMarkdown from 'react-markdown';
import { cn } from '@/utils/cn';

interface MarkdownRendererProps {
  content: string;
  className?: string;
}

export const MarkdownRenderer = memo(({ content, className }: MarkdownRendererProps) => {
  return (
    <div className={cn('prose prose-sm max-w-none dark:prose-invert', className)}>
      <ReactMarkdown
        components={{
          // Headers
          h1: ({ children }) => (
            <h1 className="text-2xl font-bold text-primary mb-4 mt-6 first:mt-0">
              {children}
            </h1>
          ),
          h2: ({ children }) => (
            <h2 className="text-xl font-semibold text-primary mb-3 mt-5">
              {children}
            </h2>
          ),
          h3: ({ children }) => (
            <h3 className="text-lg font-medium text-primary mb-2 mt-4">
              {children}
            </h3>
          ),
          
          // Paragraphs
          p: ({ children }) => (
            <p className="text-secondary mb-4 leading-relaxed">
              {children}
            </p>
          ),
          
          // Lists
          ul: ({ children }) => (
            <ul className="list-disc list-inside mb-4 space-y-1 text-secondary">
              {children}
            </ul>
          ),
          ol: ({ children }) => (
            <ol className="list-decimal list-inside mb-4 space-y-1 text-secondary">
              {children}
            </ol>
          ),
          li: ({ children }) => (
            <li className="ml-4 text-secondary">{children}</li>
          ),
          
          // Code blocks
          code: ({ className, children, ...props }: any) => {
            const inline = props.inline;
            const match = /language-(\w+)/.exec(className || '');
            
            if (!inline && match) {
              return (
                <div className="relative group mb-4">
                  <pre className="bg-background-secondary rounded-lg p-4 overflow-x-auto">
                    <code className={cn('text-sm', className)} {...props}>
                      {children}
                    </code>
                  </pre>
                </div>
              );
            }
            
            return (
              <code
                className={cn(
                  'bg-secondary/20 rounded px-1.5 py-0.5 text-sm font-mono text-primary',
                  className
                )}
                {...props}
              >
                {children}
              </code>
            );
          },
          
          // Links
          a: ({ href, children }) => (
            <a
              href={href}
              target="_blank"
              rel="noopener noreferrer"
              className="text-blue-500 hover:text-blue-600 underline"
            >
              {children}
            </a>
          ),
          
          // Blockquotes
          blockquote: ({ children }) => (
            <blockquote className="border-l-4 border-secondary/30 pl-4 italic text-secondary mb-4">
              {children}
            </blockquote>
          ),
          
          // Horizontal rules
          hr: () => <hr className="my-6 border-secondary/20" />,
          
          // Tables
          table: ({ children }) => (
            <div className="overflow-x-auto mb-4">
              <table className="min-w-full divide-y divide-secondary/20">
                {children}
              </table>
            </div>
          ),
          thead: ({ children }) => (
            <thead className="bg-secondary/10">{children}</thead>
          ),
          tbody: ({ children }) => (
            <tbody className="divide-y divide-secondary/10">{children}</tbody>
          ),
          tr: ({ children }) => <tr>{children}</tr>,
          th: ({ children }) => (
            <th className="px-3 py-2 text-left text-sm font-semibold text-primary">
              {children}
            </th>
          ),
          td: ({ children }) => (
            <td className="px-3 py-2 text-sm text-secondary">{children}</td>
          ),
          
          // Images
          img: ({ src, alt }) => (
            <img
              src={src}
              alt={alt}
              className="rounded-lg max-w-full h-auto mb-4"
            />
          ),
        }}
      >
        {content}
      </ReactMarkdown>
    </div>
  );
});

MarkdownRenderer.displayName = 'MarkdownRenderer';