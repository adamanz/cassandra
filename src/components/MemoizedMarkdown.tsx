import { Marked } from 'marked';
import { memo, useMemo } from 'react';
import ReactMarkdown from 'react-markdown';

interface MarkdownProps {
  content: string;
  id: string;
}

// Optimize the block component with a simple memo comparison
const MemoizedMarkdownBlock = memo<{ content: string }>(
  ({ content }) => <ReactMarkdown>{content}</ReactMarkdown>
);

MemoizedMarkdownBlock.displayName = 'MemoizedMarkdownBlock';

// Main component with optimized rendering
export const MemoizedMarkdown = memo<MarkdownProps>(({ content, id }) => {
  // Only recompute blocks when content changes
  const blocks = useMemo(() => {
    const marked = new Marked();
    const tokens = marked.lexer(content);
    return tokens.map(token => token.raw);
  }, [content]);

  // Return early if no blocks to render
  if (!blocks.length) return null;

  // Use array index as key since blocks are stable based on content
  return (
    <>
      {blocks.map((block, index) => (
        <MemoizedMarkdownBlock 
          content={block} 
          key={`${id}-${index}`} 
        />
      ))}
    </>
  );
});

MemoizedMarkdown.displayName = 'MemoizedMarkdown';