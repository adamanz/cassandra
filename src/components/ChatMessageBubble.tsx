import type { Message } from 'ai/react';
import { MarkdownRenderer } from './MarkdownRenderer';
import { cn } from '@/utils/cn';

export function ChatMessageBubble(props: { message: Message; aiEmoji?: string }) {
  return (
    <div
      className={cn(
        `rounded-lg max-w-[80%] mb-8 flex`,
        props.message.role === 'user' ? 'bg-secondary text-secondary-foreground px-4 py-2' : null,
        props.message.role === 'user' ? 'ml-auto' : 'mr-auto',
      )}
    >
      {props.message.role !== 'user' && (
        <div className="mr-4 bg-secondary rounded-full w-10 h-10 flex-shrink-0 flex items-center justify-center">
          {props.aiEmoji}
        </div>
      )}

      <div className="chat-message-bubble whitespace-pre-wrap flex flex-col">
        <MarkdownRenderer content={props.message.content} />
      </div>
    </div>
  );
}
