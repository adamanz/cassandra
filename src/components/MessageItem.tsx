import React, { memo } from 'react';
import { type Message } from 'ai/react';
import { MarkdownRenderer } from './MarkdownRenderer';
import { AgentThinkingLoader } from './AgentThinkingLoader';
import { cn } from '@/utils/cn';

interface MessageItemProps {
  message: Message;
  aiEmoji?: string;
  isLoading?: boolean;
}

export const MessageItem = memo(({ message, aiEmoji = 'ğŸ“…', isLoading = false }: MessageItemProps) => {
  const isUser = message.role === 'user';
  
  if (isUser) {
    return (
      <div className="flex items-start justify-end mb-6">
        <div className="flex items-start gap-3 max-w-[80%]">
          <div className="flex flex-col">
            <div className={cn(
              "rounded-xl px-4 py-2.5",
              "bg-primary text-primary-foreground",
              "shadow-sm"
            )}>
              <p className="text-sm">{message.content}</p>
            </div>
          </div>
          <div className="flex-shrink-0 w-8 h-8 rounded-full bg-primary/10 flex items-center justify-center">
            <span className="text-xs">ğŸ‘¤</span>
          </div>
        </div>
      </div>
    );
  }
  
  // AI message
  return (
    <div className="flex items-start mb-6">
      <div className="flex items-start gap-3 max-w-[80%]">
        <div className="flex-shrink-0 w-8 h-8 rounded-full bg-secondary flex items-center justify-center">
          <span className="text-sm">{aiEmoji}</span>
        </div>
        <div className="flex flex-col">
          {isLoading ? (
            <div className="rounded-xl px-4 py-3 bg-background-secondary">
              <AgentThinkingLoader />
            </div>
          ) : (
            <div className={cn(
              "rounded-xl px-4 py-3",
              "bg-background/50 dark:bg-background-secondary",
              "shadow-sm border border-border/50"
            )}>
              <MarkdownRenderer content={message.content} />
            </div>
          )}
        </div>
      </div>
    </div>
  );
});

MessageItem.displayName = 'MessageItem';