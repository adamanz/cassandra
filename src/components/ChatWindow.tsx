'use client';

import { type Message } from 'ai';
import { useChat } from '@ai-sdk/react';
import { useState, memo, useCallback, useMemo } from 'react';
import type { FormEvent, ReactNode } from 'react';
import { StickToBottom, useStickToBottomContext } from 'use-stick-to-bottom';
import { ArrowDown, ArrowUpIcon, LoaderCircle, AlertCircleIcon, RefreshCcw } from 'lucide-react';

import { ChatMessageBubble } from '@/components/ChatMessageBubble';
import { Button } from '@/components/ui/button';
import { cn } from '@/utils/cn';

const ChatMessages = memo(function ChatMessages(props: {
  messages: Message[];
  emptyStateComponent: ReactNode;
  aiEmoji?: string;
  className?: string;
}) {
  return (
    <div className="flex flex-col max-w-[768px] mx-auto pb-12 w-full">
      {props.messages.map((m) => (
        <ChatMessageBubble key={m.id} message={m} aiEmoji={props.aiEmoji} />
      ))}
    </div>
  );
});

const ScrollToBottom = memo(function ScrollToBottom(props: { className?: string }) {
  const { isAtBottom, scrollToBottom } = useStickToBottomContext();

  if (isAtBottom) return null;
  return (
    <Button variant="outline" className={props.className} onClick={scrollToBottom}>
      <ArrowDown className="w-4 h-4" />
    </Button>
  );
});

const ChatInput = memo(function ChatInput(props: {
  onSubmit: (e: FormEvent<HTMLFormElement>) => void;
  value: string;
  onChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  loading?: boolean;
  disabled?: boolean;
  placeholder?: string;
  children?: ReactNode;
  className?: string;
}) {
  const handleSubmit = useCallback((e: FormEvent<HTMLFormElement>) => {
    e.stopPropagation();
    e.preventDefault();
    props.onSubmit(e);
  }, [props.onSubmit]);

  return (
    <form
      onSubmit={handleSubmit}
      className={cn('flex w-full flex-col', props.className)}
    >
      <div className="border border-input bg-background rounded-lg flex flex-col gap-2 max-w-[768px] w-full mx-auto">
        <input
          value={props.value}
          placeholder={props.placeholder}
          onChange={props.onChange}
          disabled={props.disabled}
          className="border-none outline-none bg-transparent p-4"
        />

        <div className="flex justify-end mr-2 mb-2">
          <Button
            className="rounded-full p-1.5 h-fit border dark:border-zinc-600"
            type="submit"
            disabled={props.loading || props.disabled}
          >
            {props.loading ? <LoaderCircle className="animate-spin" /> : <ArrowUpIcon size={14} />}
          </Button>
        </div>
      </div>
    </form>
  );
});

const ErrorNotification = memo(function ErrorNotification(props: { 
  error: Error | null;
  onRetry: () => void;
}) {
  if (!props.error) return null;
  
  return (
    <div className="bg-destructive/10 text-destructive border border-destructive/20 rounded-lg p-3 mb-4 max-w-[768px] mx-auto w-full flex items-center justify-between">
      <div className="flex items-center gap-2">
        <AlertCircleIcon className="w-4 h-4" />
        <p className="text-sm">Error connecting. Try logging out and back in.</p>
      </div>
      <Button variant="outline" size="sm" onClick={props.onRetry} className="gap-1 px-2 py-1 h-auto">
        <RefreshCcw className="w-3 h-3" />
      </Button>
    </div>
  );
});

const StickyToBottomContent = memo(function StickyToBottomContent(props: {
  content: ReactNode;
  footer?: ReactNode;
  className?: string;
  contentClassName?: string;
}) {
  const context = useStickToBottomContext();

  return (
    <div
      ref={context.scrollRef}
      style={{ width: '100%', height: '100%' }}
      className={cn('grid grid-rows-[1fr,auto]', props.className)}
    >
      <div ref={context.contentRef} className={props.contentClassName}>
        {props.content}
      </div>
      {props.footer}
    </div>
  );
});

export const ChatWindow = memo(function ChatWindow(props: {
  endpoint: string;
  emptyStateComponent: ReactNode;
  placeholder?: string;
  emoji?: string;
}) {
  const [hasError, setHasError] = useState<Error | null>(null);
  
  const chat = useChat({
    api: props.endpoint,
    onFinish() {
      setHasError(null);
    },
    onResponse(response) {
      if (!response.ok) {
        response.json().then((data) => {
          setHasError(new Error(data.error || 'Unknown error occurred'));
        }).catch(() => {
          setHasError(new Error(`Request failed with status ${response.status}`));
        });
      } else {
        setHasError(null);
      }
    },
    onError: (e) => {
      setHasError(e);
    },
  });

  const isChatLoading = useCallback((): boolean => {
    return chat.status === 'streaming';
  }, [chat.status]);

  const sendMessage = useCallback(async (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    if (isChatLoading()) return;
    setHasError(null);
    chat.handleSubmit(e);
  }, [chat, isChatLoading]);
  
  const retryConnection = useCallback(() => {
    setHasError(null);
    window.location.href = "/auth/logout?returnTo=/";
  }, []);

  const isLoading = isChatLoading();
  const placeholder = useMemo(() => {
    return hasError 
      ? 'Please retry connection before sending messages' 
      : props.placeholder ?? 'What can I help you with?';
  }, [hasError, props.placeholder]);

  const content = useMemo(() => {
    if (chat.messages.length === 0) {
      return <div>{props.emptyStateComponent}</div>;
    }
    
    return (
      <>
        {hasError && <ErrorNotification error={hasError} onRetry={retryConnection} />}
        <ChatMessages
          aiEmoji={props.emoji}
          messages={chat.messages}
          emptyStateComponent={props.emptyStateComponent}
        />
      </>
    );
  }, [chat.messages, hasError, props.emoji, props.emptyStateComponent, retryConnection]);

  const footer = useMemo(() => (
    <div className="sticky bottom-8 px-2">
      <ScrollToBottom className="absolute bottom-full left-1/2 -translate-x-1/2 mb-4" />
      <ChatInput
        value={chat.input}
        onChange={chat.handleInputChange}
        onSubmit={sendMessage}
        loading={isLoading}
        disabled={!!hasError}
        placeholder={placeholder}
      />
    </div>
  ), [chat.input, chat.handleInputChange, sendMessage, isLoading, hasError, placeholder]);

  return (
    <StickToBottom>
      <StickyToBottomContent
        className="absolute inset-0"
        contentClassName="py-8 px-2"
        content={content}
        footer={footer}
      />
    </StickToBottom>
  );
});
