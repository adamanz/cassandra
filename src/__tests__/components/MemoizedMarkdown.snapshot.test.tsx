import React from 'react';
import { render } from '@testing-library/react';
import { MemoizedMarkdown } from '../../components/MemoizedMarkdown';
import { Marked } from 'marked';

// Mock the marked library with consistent output
jest.mock('marked', () => {
  return {
    Marked: jest.fn().mockImplementation(() => {
      return {
        lexer: jest.fn((text) => {
          // Return consistent tokens for snapshot testing
          if (text.includes('# Header')) {
            return [
              { raw: '# Header\n', type: 'heading', depth: 1, text: 'Header' },
              { raw: 'This is a paragraph.\n', type: 'paragraph', text: 'This is a paragraph.' },
              { raw: '- Item 1\n', type: 'list', ordered: false, items: ['Item 1'] },
              { raw: '- Item 2\n', type: 'list', ordered: false, items: ['Item 2'] }
            ];
          }
          return text.split('\n').map((line, index) => ({
            raw: line + '\n',
            type: 'paragraph',
            text: line
          }));
        })
      };
    })
  };
});

// Mock ReactMarkdown to render predictable output
jest.mock('react-markdown', () => {
  return function MockReactMarkdown({ children }: { children: string }) {
    // Parse the raw markdown for snapshot consistency
    const trimmed = children.trim();
    
    if (trimmed.startsWith('# ')) {
      return <h1>{trimmed.substring(2)}</h1>;
    } else if (trimmed.startsWith('- ')) {
      return <li>{trimmed.substring(2)}</li>;
    } else if (trimmed) {
      return <p>{trimmed}</p>;
    }
    
    return null;
  };
});

describe('MemoizedMarkdown Snapshots', () => {
  it('should render simple markdown correctly', () => {
    const { container } = render(
      <MemoizedMarkdown 
        content="Hello world\nThis is a test" 
        id="simple-snapshot" 
      />
    );
    
    expect(container).toMatchSnapshot();
  });

  it('should render complex markdown with headers and lists', () => {
    const complexContent = `# Header
This is a paragraph.
- Item 1
- Item 2`;
    
    const { container } = render(
      <MemoizedMarkdown 
        content={complexContent} 
        id="complex-snapshot" 
      />
    );
    
    expect(container).toMatchSnapshot();
  });

  it('should render empty content', () => {
    const { container } = render(
      <MemoizedMarkdown 
        content="" 
        id="empty-snapshot" 
      />
    );
    
    expect(container).toMatchSnapshot();
  });

  it('should render with different IDs', () => {
    const content = "Test content";
    
    const { container: container1 } = render(
      <MemoizedMarkdown content={content} id="id-1" />
    );
    
    const { container: container2 } = render(
      <MemoizedMarkdown content={content} id="id-2" />
    );
    
    expect(container1).toMatchSnapshot('with-id-1');
    expect(container2).toMatchSnapshot('with-id-2');
  });
});