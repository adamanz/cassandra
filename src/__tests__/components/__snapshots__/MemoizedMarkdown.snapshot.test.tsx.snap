// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemoizedMarkdown Snapshots should render complex markdown with headers and lists 1`] = `
<div>
  <h1>
    Header
  </h1>
  <p>
    This is a paragraph.
  </p>
  <li>
    Item 1
  </li>
  <li>
    Item 2
  </li>
</div>
`;

exports[`MemoizedMarkdown Snapshots should render empty content 1`] = `<div />`;

exports[`MemoizedMarkdown Snapshots should render simple markdown correctly 1`] = `
<div>
  <p>
    Hello world\\nThis is a test
  </p>
</div>
`;

exports[`MemoizedMarkdown Snapshots should render with different IDs: with-id-1 1`] = `
<div>
  <p>
    Test content
  </p>
</div>
`;

exports[`MemoizedMarkdown Snapshots should render with different IDs: with-id-2 1`] = `
<div>
  <p>
    Test content
  </p>
</div>
`;
